### DISCUSSION TITLE
Separate Python virtual environments for each DAG(not Task)

### DISCUSSION THREAD

**Discussion by nimanov** [apache/airflow#40651] [Q&A]
# Separate virtual environments for each DAG

Dear community,

Is there a way to declare the virtual environment(that is already created) the DAG will use for all the tasks that it has.

I am aware of the `ExternalPythonOperator` however it is used for each task which is kinda tedious to write for each task if there is a way to do only once for DAG. 

Do we have some way to do that? 

If not, could you please explain why this isn't possible? Is it because each task runs as a separate process by the worker?
I would appreciate a detailed answer if the answer is 'no'.

This need is required when 2 or more teams/developers  working to create DAGs. They may use different versions of packages for whole DAG. Thus, declaring that the DAG will use that environment for all the tasks would be nice feature to have.

Thanks in advance.

**Comment by potiuk** [comment]
No.

Mainly because no-one contributed it.

But since you need it and would like to use it, I would appreciate if you and your company would contribute such feature - similarly as nearly 3000 contributors so far - many of whom went the same route - they needed something and contributed it.

There are certain limitations when you want to "change" interpreter on the flight - mainly due to the fact that you need to serialize and deserialize input parameters and output values (this is what PythonVirtualenv and  ExternalPython operators both do) - you need to pass them from one interpreter to another. So not all Python code will run because not all arguments and return values can be serialized.

But you have also plenty of other options - you can use default_args at the DAG level to specify virtualenv each of your ExternalPythonOperators will use. You can build your own custom operators extending from ExternalPyhon operator (separately for each team) or you can have each team use different queues - as described in the best practices (see last point in the list https://airflow.apache.org/docs/apache-airflow/stable/best-practices.html#handling-conflicting-complex-python-dependencies) - so the teams are separate on deployment level, you can even use cluster policies to force the queue.

Finally in Airflow 3 there is a multi-team feature planned https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-67+Multi-team+deployment+of+Airflow+components that you will be able to utilise to separate your team workfloads.

But in the meantime if you have an idea how it shoudl look like to specify different virtual environment for DAG to utilize for all tasks - you are absolutely welcome to contribute such feature. You seem to have quite a big team so it might be easy for you to spare some engineering power to do so. You are absolutely welcome to contribute it, That would be a nice way for your company to give back for the software you got absolutely for free  - and this is how it works for others who contributed.

Can we count on you ?

**Comment by GShirey** [comment]
What I have done in the past is to create a custom operator extending from the bash operator and utilized the tags value that we provided in the dag definition/configuration. Depending on the value of the tag, then it would source the correct virtual environment before running the original bash command. The tag value could be each team/area that would need it's own virtual env.

tags = context['dag'].tags

if tags:
            for tag in tags:
                if "xxxxxxxx" in tags:
                    virtual_env_activate = f"source /virtualenv/{tag}/bin/activate"
            bash_command_new = f'{virtual_env_activate} && {self.bash_command}        
self.bash_command=bash_command_new

**Comment by HariNuve** [comment]
I also have similar issue,some of the modules related to a particular dag not supports current environmnet that iam working,i need to findout a solution to encorporate multiple environments,Could anyone can suggest me any ideas to overcome this issue??all dags except one do not working in current environment due to a particular dependency conflict,that why i have created separate environment for the working of that scrypt.

**Comment by pykenny** [comment]
I read the [slides about remote worker](https://airflowsummit.org/sessions/2024/running-airflow-tasks-anywhere-in-any-language/) (or edge worker, to be more specifically) introduced on last year's Airflow Summit and feel that it could possibly be the solution when the feature becomes fully available. An executor still need to run in an environment where the same Python+Airflow version as the core service does, but at runner level it won't matter as long as it supports the recognized task execution protocol (using compatible Task SDK to communicate with executor). Finally DAG authors can write task flow in this way to specify the environment(s):

```python
@dag(...):
def my_dag:
   # ...
   task_venv = task.external("environment_001")
   upstream = task_venv.upstream_job()
   task_venv.downstream_job(data=upstream)
   # ...
```

And on the machine where the edge executor runs, runners will have access to the following task definitions:

```python
from airflow.sdk import task  # Just Task SDK installed on the environment, Core not included

@task(task_id="upstream_job", **context)
def upstream_job() -> str:
    # ...

@task(task_id="downstream_job", **context)
def downstream_job(data):
    # ...
```

Unfortunately at the time of writing, Airflow officials haven't release any SDK that supports language outside of Python (Golang, JavaScript, TypeScript, etc.) yet, and current Python Task SDK has to take care of [backward compatibility of existing codebase](https://airflow.apache.org/blog/airflow-three-point-oh-is-here/#:~:text=The%20Airflow%203%20release%20includes%20the%20Python%20TaskSDK%20which%20enables%20backward%20compatibility%20for%20existing%20DAGs.%20TaskSDKs%20for%20additional%20languages%2C%20starting%20with%20Golang%20will%20be%20released%20over%20the%20next%20few%20months.), meaning it's [heavily tied to the core system itself](https://github.com/apache/airflow/blob/35186bccf3da10d5d01b744966be8156721be5c7/task-sdk/pyproject.toml#L49) (runner using Python SDK should run in an environment with Airflow installed), which doesn't offer the same flexibility as what's demonstrated in the summit.

**Reply by pykenny** [reply]
Otherwise, [as earlier response](https://github.com/apache/airflow/discussions/40651#discussioncomment-9988010) suggested, one may need to tackle with remote executors (Kubernetes or Celery) to deploy remote workers using different Python environments (of course, both requires Airflow to be available in worker's runtime environment).

**Reply by pykenny** [reply]
And in the long run, when Task SDK (what AIP-72 does) becomes independent from the core utilities and AIP-67 (multi-team deployment) completes, developers will be able to define task script or even DAG definition that runs on their managed environment.
